File: C#/RefactorMe/ProductDataConsolidator.cs
using System;
using RefactorMe.DontRefactor.Data.Implementation;
using RefactorMe.DontRefactor.Models;
using System.Collections.Generic;
using System.Linq;

namespace RefactorMe
{
	public interface IProduct
	{
		int ID { get; set; }
		string Name { get; set; }
		double Price { get; set; }
		string Type { get; set; }
	}

	public class ProductDataConsolidator
    {
		public ProductTypeBuilder ProductTypeBuilder { get; set; } = new ProductTypeBuilder();

		private List<Product> ConsolidatedProducts { get; set; } = new List<Product>();

		/*private void _AddConsolidatedProduct(Guid id, string name, double price, string type)
		{
			this.ConsolidatedProducts.Add(new Product {Id = id, Name = name, Price = price, Type = type});
		}*/

		private void _AddConsolidatedProduct(IProduct product, string type)
		{
			Console.WriteLine(product.ToString());
		}

		/// <summary>
		/// Returns a list of consolidated products.
		/// </summary>
		/// <param name="unconsolidatedProducts"></param>
		/// <returns></returns>
		public List<Product> GetProducts(List<object> unconsolidatedProducts)
		{
			unconsolidatedProducts.ForEach(p =>
			{
				var type = ProductTypeBuilder.GetTypeString(p);
				p = (IProduct) p;

				switch (type)
				{
					case "Lawnmower":
						var l = (IProduct) p;
						this._AddConsolidatedProduct(l, type);
						break;
					case "T-Shirt":
						var ts = (IProduct) p;
						this._AddConsolidatedProduct(ts, type);
						break;
					case "Phone Case":
						this._AddConsolidatedProduct((IProduct) p, type);
						break;
					default:
						throw new NotSupportedException("The product type is not supported");
				}
			});
			return this.ConsolidatedProducts;
			/*
		    return new LawnmowerRepository().GetAll().Select(l => new Product
			    {Id = l.Id, Name = l.Name, Price = l.Price, Type = ProductTypeBuilder.GetTypeString(l)})
		    .Concat(new PhoneCaseRepository().GetAll().Select(pc => new Product
				{Id = pc.Id, Name = pc.Name, Price = pc.Price, Type = ProductTypeBuilder.GetTypeString(pc)})
			.Concat(new TShirtRepository().GetAll().Select(ts => new Product
				{Id = ts.Id, Name = ts.Name, Price = ts.Price, Type = ProductTypeBuilder.GetTypeString(ts)}))).ToList();*/
		}
    }
}

File: JavaScript/src/refactor-me/ProductDataConsolidator.js
/**
 * A class for for formatting consolidated products.
 */
class Product {
	constructor(id, name, price, type){
		this.id = id;
		this.name = name;
		this.price = price
		this.type = type;
	}

	/**
	 * converts product price using the given currency exchange rate.
	 * @param {exchangeRate} currency exchange rate
	 * @returns {convertedProduct} product with converted price.
	 */
	ConvertCurrency = function (exchangeRate){
		let convertedProduct = new Product(this.id, this.name, (this.price * exchangeRate).toFixed(2), this.type);
		return convertedProduct;
	}
}

/**
 * Product Data Consolidator main method.
 */
function ProductDataConsolidator() { }

/**
 * Returns a list of consolidated products with their price in default currency (New Zealand Dollar)
 * @returns {products} list of consolidated products
 */
ProductDataConsolidator.get = function () {
	let lawnmowers = new LawnmowerRepository().getAll();
	let phoneCases = new PhoneCaseRepository().getAll();
	let tShirts = new TShirtRepository().getAll();

	let products = [];
	
	lawnmowers.forEach(l => products.push(new Product(l.id, l.name, l.price.toFixed(2), "Lawnmower")));
	phoneCases.forEach(p=> products.push(new Product(p.id, p.name, p.price.toFixed(2), "Phone Case")));
	tShirts.forEach(t=> products.push(new Product(t.id, t.name, t.price.toFixed(2), "T-Shirt")));

	return products;
}

// Returns a list of consolidated products with their price converted to US Dollar.
/**
 * @returns {products} list of consolidated products with their currency converted to US Dollar.
 */
ProductDataConsolidator.getInUSDollars = function () {
	let products = ProductDataConsolidator.get().map(p => p.ConvertCurrency(0.76));
	return products
}


// Returns a list of consolidated products with their prices converted to Euros
/**
 * 
 * @returns {products} list of consolidated products with their currency converted to Euro.
 */
ProductDataConsolidator.getInEuros = function () {
	let products = ProductDataConsolidator.get().map(p => p.ConvertCurrency(0.67));
	return products;
}